Optimized search and json deserialization

ashpatil-msft opened the pull request:
This PR contains the following updates:
- Optimized search method that will get triggered when user wants to install a package with specific params.
- Deserializers for Rest API responses for manifest, search response and information APIs.
- Basic call to /information endpoint to get supported version.

Todo:
- Will be adding unit tests and E2E tests in the upcoming PRs for the above changes.
- Implementing Search Request usage in rest source search method and version decision logic.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/800)

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/1_0/Interface.cpp:
```diff
@@ -89,115 +78,115 @@ namespace AppInstaller::Repository::Rest::Schema::V1_0
+                if (errors > 0)
+                {
+                    AICLI_LOG(Repo, Verbose, << "Received invalid manifest. Skipping");
+                    return {};
```
Should throw.

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/1_0/Interface.cpp:
```diff
@@ -89,115 +78,115 @@ namespace AppInstaller::Repository::Rest::Schema::V1_0
-        HttpClientHelper clientHelper{ GetManifestByVersionEndpoint(m_restApiUri, packageId, version, channel) };
-        web::json::value jsonObject = clientHelper.HandleGet();
+        SearchResult searchResult;
+        std::optional<Manifest::Manifest> manifestResult = GetManifestByVersion(request.Filters[0].Value, {}, {});
```
I thought that this API would return *ALL* manifests if no version was supplied.

ashpatil-msft replied:
you are correct. I will add a method that gets all the returned versions from the endpoint and will have methods call that and decide what to do with the multiple versions case by case.

---
In reply to: [596382444](https://github.com/microsoft/winget-cli/pull/800#discussion_r596382444) [](ancestors = 596382444)

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/Json/InformationResponseDeserializer.cpp:
```diff
@@ -0,0 +1,90 @@
+        if (!data.has_value())
+        {
+            AICLI_LOG(Repo, Verbose, << "Missing data");
+            return {};
```
General error handling strategy for these deserialization methods:
1. Log as Error
2. Returning an empty optional is fine *as long as* the caller treats that as an error and throws an appropriate error understandable to our stack.
3. Wrap in a function level try/catch that also returns an empty optional.
4. I would suggest combining 2 and 3 by making the public method non-optional and throw and a private one that does the work.

JohnMcPMS replied:
Basically, don't ignore error conditions by just skipping them and also don't propagate the REST specific exception types out of the source.

If a value is truly optional, then you wouldn't return early.

ashpatil-msft replied:
Have added the errors and made public methods throw when it gets an optional value.

---
In reply to: [596389291](https://github.com/microsoft/winget-cli/pull/800#discussion_r596389291) [](ancestors = 596389291)

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/Json/InformationResponseDeserializer.cpp:
```diff
@@ -0,0 +1,90 @@
+        constexpr std::string_view SourceIdentifier = "SourceIdentifier"sv;
+        constexpr std::string_view ServerSupportedVersions = "ServerSupportedVersions"sv;
+        constexpr std::string_view APIVersion = "APIVersion"sv;
+        constexpr std::string_view ObjectVersions = "ObjectVersions"sv;
```
I don't understand what all of these versions are for.

ashpatil-msft replied:
Updated based on offline conversation.

---
In reply to: [596390423](https://github.com/microsoft/winget-cli/pull/800#discussion_r596390423) [](ancestors = 596390423)

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/Json/SearchResponseDeserializer.cpp:
```diff
@@ -0,0 +1,98 @@
+
+                if (!JsonHelper::IsValidNonEmptyStringValue(packageId) || !JsonHelper::IsValidNonEmptyStringValue(packageName) || !JsonHelper::IsValidNonEmptyStringValue(packageName))
+                {
+                    AICLI_LOG(Repo, Verbose, << "Skipping manifest item because of missing required package fields.");
```
We should fail if required values are missing.

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/Json/SearchResponseDeserializer.cpp:
```diff
@@ -0,0 +1,98 @@
+        }
+        catch (...)
+        {
+            AICLI_LOG(Repo, Verbose, << "Error encountered while deserializing search result...");
```
TODO: Catch known types and log error information from them

JohnMcPMS commented on src/AppInstallerRepositoryCore/Rest/Schema/Json/ManifestDeserializer.cpp:
```diff
@@ -0,0 +1,397 @@
+    namespace
+    {
+        // Manifest response constants specific to this deserializer
+        constexpr std::string_view PackageIdentifier = "PackageIdentifier"sv;
```
Proof that having the tree walking interface and single definition of the manifest parser would be much less code.  This would literally not need to know anything about the manifest structure (assuming that we actually defined the same tree for the REST response... please tell me we did that).

JohnMcPMS replied:
Not heavily reviewed because I think we should do that work sooner rather than later.

ashpatil-msft commented on src/AppInstallerRepositoryCore/Rest/Schema/1_0/Interface.cpp:
```diff
@@ -89,115 +78,115 @@ namespace AppInstaller::Repository::Rest::Schema::V1_0
     }
 
-    Interface::Interface(const std::string& restApi)
+    Interface::Interface(const std::string& restApi, const std::string& restApiVersion)
```
This was supposed to be the rest API Version, but now that we have removed it from our schema, I will remove this.

---
In reply to: [596376550](https://github.com/microsoft/winget-cli/pull/800#discussion_r596376550) [](ancestors = 596376550)

ashpatil-msft commented on src/AppInstallerCommonCore/Errors.cpp:
```diff
@@ -125,6 +125,10 @@ namespace AppInstaller
+            case APPINSTALLER_CLI_ERROR_UNSUPPORTED_RESTSOURCE:
+                return "The configured rest source is not supported";
+            case APPINSTALLER_CLI_ERROR_RESTSOURCE_INVALID_DATA:
+                return "Invalid data returned by rest source";
```
Let me know if you want to change the wording/name of these errors

JohnMcPMS reviewed (approved):
A few minor things.

ashpatil-msft commented:
Addressed the comments..

---
In reply to: [616735978](https://github.com/microsoft/winget-cli/pull/800#issuecomment-616735978) [](ancestors = 616735978)
